src/coreclr/pal/src/map/virtual.cpp

```diff
@@ -2136,11 +2138,33 @@ void ExecutableMemoryAllocator::Initialize()
 void ExecutableMemoryAllocator::TryReserveInitialMemory()
 {
     CPalThread* pthrCurrent = InternalGetCurrentThread();
-    int32_t sizeOfAllocation = MaxExecutableMemorySizeNearCoreClr;
     int32_t preferredStartAddressIncrement;
     UINT_PTR preferredStartAddress;
     UINT_PTR coreclrLoadAddress;
 
+    int32_t sizeOfAllocation = MaxExecutableMemorySizeNearCoreClr;
+    int32_t initialReserveLimit = -1;
+    rlimit addressSpaceLimit;
+    if ((getrlimit(RLIMIT_AS, &addressSpaceLimit) == 0) && (addressSpaceLimit.rlim_cur != RLIM_INFINITY))
+    {
+        // By default reserve max 25% of the available virtual address space
+        rlim_t initialExecMemoryPerc = 25;
+        CLRConfigNoCache defInitialExecMemoryPerc = CLRConfigNoCache::Get("InitialExecMemoryPercent", /*noprefix*/ false, &getenv);
+        if (defInitialExecMemoryPerc.IsSet())
+        {
+            DWORD perc;
+            if (defInitialExecMemoryPerc.TryAsInteger(16, perc))
```

Comment:
It feels a bit strange to specify the percentage in hex, but it is the default we use for all COMPlus / DOTNET variables, including percentual settings in the GC.

Reply:
I don't know if it is architecturally feasible, but it would be nice if we don't need to proliferate the environment variable parsing rules.